test/plugins/adonis.spec.js

```diff
@@ -0,0 +1,386 @@
+        const randId = Math.ceil(Math.random() * 100)
+
+        class AppMiddleware {
+          handle (ctx, next, [id]) {
+            return new Promise((resolve, reject) => {
+              resolve(ctx)
+            }).then(next)
+              .catch((error) => {})
+          }
+        }
+
+        if (semver.intersects(adonisVersion, '>=5.0.3')) {
+          before(() => {
+            const framework = require(`${frameworkPath}@${adonisVersion}`)
+
+            Server = framework.get('@adonisjs/framework/src/Server')
+            Route = framework.get('@adonisjs/framework/src/Route/Manager')
+            RouteStore = framework.get('@adonisjs/framework/src/Route/Store')
+            Request = framework.get('@adonisjs/framework/src/Request')
+            Response = framework.get('@adonisjs/framework/src/Response')
+            Context = framework.get('@adonisjs/framework/src/Context')
+            Exception = framework.get('@adonisjs/framework/src/Exception')
+            MiddlewareBase = require(`${middlewarePath}@${middlewareVersion}`).get()
+
+            Context.getter('response', function () {
+              return new Response(this.request, new Config())
+            }, true)
+
+            ioc.autoload(path.join(__dirname), 'App')
+          })
+        } else {
+          before(() => {
+            Server = require(`${frameworkPath}@${adonisVersion}`).get('@adonisjs/framework/src/Server')
+            Route = require(`${frameworkPath}@${adonisVersion}`).get('@adonisjs/framework/src/Route/Manager')
+            RouteStore = require(`${frameworkPath}@${adonisVersion}`).get('@adonisjs/framework/src/Route/Store')
+            Request = require(`${frameworkPath}@${adonisVersion}`).get('@adonisjs/framework/src/Request')
+            Response = require(`${frameworkPath}@${adonisVersion}`).get('@adonisjs/framework/src/Response')
+            Context = require(`${frameworkPath}@${adonisVersion}`).get('@adonisjs/framework/src/Context')
+            Exception = require(`${frameworkPath}@${adonisVersion}`).get('@adonisjs/framework/src/Exception')
+            MiddlewareBase = require(`${middlewarePath}@${middlewareVersion}`).get()
+
+            Context.getter('response', function () {
+              return new Response(this.req, this.res, new Config())
+            }, true)
+          })
+        }
+
+        before(() => {
+          Context.getter('request', function () {
+            return new Request(this.req, this.res, new Config())
+          }, true)
+
+          setupResolver()
+          agent.load(plugin, 'adonis').then(() => {
+          })
+        })
+
+        beforeEach(() => {
+          this.exception = Exception
+          this.server = new Server(Context, Route, new Logger(), this.exception)
+          this.middleware = new MiddlewareBase('handle')
+
+          getPort().then(newPort => {
+            port = newPort
+            server = this.server.listen('localhost', port)
+            server.on('listening', () => {
```

Comment:
I think you want to use `done` here as the callback, and still follow the pattern of making the requests within the test after route/middleware registrations and declaring the assertions that's used in all framework tests.

Reply:
Could you please explain why I'll want to use `done` as a call back? Also, are you referring to having the `server.listen` function in each test call instead of before each?